### PR TITLE
Make `impl<Fd: AsFd>` impl take `?Sized`

### PR DESCRIPTION
**Pull request by nbdd0121**
@rustbot labels: +T-libs-api +needs-fcp

### CODE REVIEW

**Comment by nbdd0121** [comment]
As discussed in https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/!Sized.20AsFd, since `&, &mut, Box` are fundamental types and changing the 3 relevant blanket impl is technically breaking. So this needs a crater run.

**Comment by Dylan-DPC** [comment]
@craterbot run

**Comment by Dylan-DPC** [comment]
@craterbot check

**Comment by dtolnay** [comment]
@craterbot check

**Comment by dtolnay** [comment]
Crater claims 6 regressions. 5 of them are almost certainly unrelated to this PR. The one in gh/lcdr.world_server looks _maybe_ related, but I tried to reproduce it locally using `cargo check --frozen --all --all-targets` (which is the command run by Crater) and it compiles equally with this PR as with nightly. I tried both with this PR as is, and this PR rebased onto current master, and neither reproduces the failure from crater.

In conclusion, _if_ there is any breakage, it would be extremely uncommon.

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

**Comment by dtolnay** [comment]
@nbdd0121 we should consider making an equivalent change for the Windows `AsHandle` impls too.

**Comment by taiki-e** [comment]
Is there any reason why this is done only for AsFd and AsHandle but not for AsRawFd and AsSocket?

https://github.com/rust-lang/rust/blob/f4aeb7030993f5c143e9a84c50061892f6b7046f/library/std/src/os/fd/raw.rs#L253
https://github.com/rust-lang/rust/blob/f4aeb7030993f5c143e9a84c50061892f6b7046f/library/std/src/os/windows/io/socket.rs#L270

**Comment by nbdd0121** [comment]
They have been in the library for so long that I am a little bit more worried to add it. If we want to add it then it should probably go in another PR and redo a crater run.

**Review by dtolnay — APPROVED** [review]
Thank you!
